Allow `type=` statements in `vfolder-from-query`

Austin-Ray opened the pull request:
**What does this PR do?**

This addresses an issue with how Neomutt encodes notmuch queries. If one uses `vfolder-from-query` and includes a `type=` statement, Neomutt will encode the statement as part of the query, which prevents the query type from being changed. Including `type=` statements is documented as supported behavior[0].

This PR achieves the desired result by introducing a case-insensitive implementation of `strstr(...)`, a function for removing all instances of a case-insensitive substring, and combining the two functions to find if a `type=` statement exists and to set the appropriate `query_type` in `NmCtxData`if one exists.

[0] https://neomutt.org/feature/notmuch#3-2-%C2%A0items

**What are the relevant issue numbers?**
#1214

flatcap commented on mutt/string.c:
```diff
@@ -959,3 +959,71 @@ const char *mutt_str_getenv(const char *name)
+{
+  // There's no point in removing something that is not there.
+  if (mutt_str_strcasestr(str, target) == NULL)
+    return 1;
```
Do we want to do this (expensive) check unnecessarily?
We might as well let the `while` do it.

flatcap commented on mutt/string.c:
```diff
@@ -959,3 +959,71 @@ const char *mutt_str_getenv(const char *name)
+  while ((str = (char *) mutt_str_strcasestr(str, target)))
+  {
+    size_t target_len = mutt_str_strlen(target);
+    memmove(str, str + target_len, 1 + strlen(str + target_len));
```
... and set a flag here, to determine the return value

Austin-Ray replied:
I am unsure how to achieve this due to my lack of familiarity with C. Could you provide an example or link to an example?

Austin-Ray replied:
Okay, I think I might have figured out how to achieve this.

Would it be something like:

```C
int retval = 0;

while ((retval = (str = (char *) mutt_str_strcasestr(str, target))))
{
   // loop body
}

return retval;
```

That seems mess-y and hard to read, but compiles and works; however, I believe it is abuse of the standard (and some poor, future individual's sanity.)

flatcap commented on notmuch/mutt_notmuch.c:
```diff
@@ -1883,6 +1883,34 @@ int nm_read_entire_thread(struct Context *ctx, struct Header *h)
+    "type=threads&", "type=messages&",
+    "type=threads", "type=messages"};
+
+  for (int i = 0; i < 6; i++)
```
Replace `6` with `mutt_array_size(variants)`

flatcap commented on notmuch/mutt_notmuch.c:
```diff
@@ -1883,6 +1883,34 @@ int nm_read_entire_thread(struct Context *ctx, struct Header *h)
+  // The six variations of how type= could appear.
+  const char *variants[6] = {"&type=threads", "&type=messages",
+    "type=threads&", "type=messages&",
+    "type=threads", "type=messages"};
```
This seems a bit awkward...
The first two are only correct if at the end of a query.
The second two, only at the start.
The final two only if they are the entire query.

Otherwise, respectively...

- `notmuch://?query=hello&type=threadsJUNK`
- `notmuch://?JUNKtype=threads&query=hello`
- `notmuch://?JUNKtype=threadsJUNK`

flatcap replied:
ok, thanks

Austin-Ray commented:
> However a couple of the tests don't work!

Which tests do not work? All test cases passed in Travis-CI.
Is it how I wrote the tests?

flatcap commented:
```
test/neomutt-test -v test_string_strcasestr

Test test_string_strcasestr:
  test/string.c:129: Check Expected: (null), Actual (null)... ok
  test/string.c:130: Check Expected: (null), Actual (null)... ok
  test/string.c:131: Check Expected: (null), Actual (null)... ok
  test/string.c:139: Check Expected: , Actual ... ok
  test/string.c:140: Check Expected: (null), Actual (null)... ok
  test/string.c:141: Check Expected: hello, Actual hello... ok
  test/string.c:149: Check Expected: (null), Actual (null)... ok
  test/string.c:150: Check Expected: (null), Actual (null)... ok
  test/string.c:151: Check Expected: (null), Actual (null)... ok
  test/string.c:157: Check Expected: hello, Actual: hello... ok
  test/string.c:163: Check Expected: hello, world!, Actual: hello, world!... ok
  test/string.c:170: Check Expected: hello, world, Actual: hello, world... failed
  test/string.c:177: Check Expected: hello, Actual: hello... failed
  test/string.c:183: Check Expected: (null), Actual: (null)... ok
  2 conditions have FAILED.

Summary:
  Count of all unit tests:       14
  Count of run unit tests:        1
  Count of failed unit tests:     1
  Count of skipped unit tests:   13
  FAILED: 1 of 1 unit tests have failed.
```

I can investigate later.

flatcap commented:
```sh
# gcc fails
gcc --version
gcc (GCC) 8.1.1 20180712 (Red Hat 8.1.1-5)

# clang works
clang --version
clang version 6.0.1 (tags/RELEASE_601/final)
```

Hmm... this is very interesting.  It's to do with optimisations.
The test at line 170 compares two pointers, rather than doing `strcmp()`.

```c
if (retval == expected)
```

In my gcc, you have two strings in the code:
- `char *haystack = "test! hello, world";`
- `const char *expected = "hello, world";`

Your compiler has noted that despite being non-const, that you don't actually modify the haystack.
It also notes that the `expected` string is a substring of `haystack`.

Effectively, it's compiling:
- `const char *haystack = "test! hello, world";`
- `const char *expected = haystack + 6;`

When you search for the `needle`, you get a pointer into the `haystack`, which coincidentally is where `expected` is.  So, `(retval == expected)` !

If the compiler doesn't make these heavy optimisations, then `(retval == expected)` will **never** be true because they're entirely separate strings.

Austin-Ray commented:
> If the compiler doesn't make these heavy optimisations, then (retval == expected) will never be true because they're entirely separate strings.

That makes complete sense. I would need to use `strcmp(...)` to actually check the string content. I have yet to grok the C way of handling strings. This error was introduced by muscle memory since I use other languages where such a comparison is valid.

Since I wrote `mutt_str_strcasestr(...)` to mimic `strstr(...)`'s behavior, would it be a good idea to use `strstr(...)` to check the output (e.g. since I know the input, i can write the `strstr(...)` with the right case then compare pointers.)?

flatcap commented:
> I use other languages where such a comparison is valid.

Yeah, we're not using `std::string`... yet.

> would it be a good idea to use strstr(...) to check the output
> ... then compare pointers

Yes, that would be good.  Check exactly what we're expecting to have returned.

Austin-Ray commented:
I addressed the test case issue, removed a redundant `mutt_str_strcasestr(...)` call, and removed the magic number from `parse_nm_query_type_from_query(...)`.

Austin-Ray commented:
@flatcap Are you still interested in this PR? I can rebase this PR if so.

flatcap commented:
> Are you still interested in this PR?
> I can rebase this PR if so.

Yes and yes please.

Austin-Ray commented:
@flatcap Branch has been rebased and passed in CI.
